### PR TITLE
Feature/referee

### PR DESCRIPTION
**Pull request by zong09**

### CODE REVIEW

**Review by julapong-castcle — APPROVED** [review]

File: apps/users/src/app/app.controller.ts [hunk 1]
```diff
@@ -831,4 +839,148 @@ export class UserController {
...
+  @Get(':id/referee')
+  @UsePipes(new ValidationPipe({ skipMissingProperties: true }))
+  async getReferee(
+    @Param('id') id: string,
+    @Req() { $credential }: CredentialRequest,
+    @Query('') userQuery: ExpansionQuery,
+    @Query('maxResults')
+    maxResults: number = DEFAULT_QUERY_OPTIONS.limit,
+    @Query('sinceId') sinceId?: string,
+    @Query('untilId') untilId?: string
```
Comment (kittipong-castcle) [inline apps/users/src/app/app.controller.ts:925]:
```suggestion
    @Query() { hasRelationshipExpansion, maxResult, sinceId, untilId }: FeedQuery
```

File: apps/users/src/app/app.controller.ts [hunk 2]
```diff
@@ -831,4 +839,148 @@ export class UserController {
...
+  ): Promise<UserRefereeResponse> {
+    if (maxResults) {
+      logger.log('validate min & max maxResults');
+      if (+maxResults < 5 || +maxResults > 100) {
+        throw new CastcleException(CastcleStatus.INVALID_MAX_RESULT);
+      }
+    }
```
Comment (kittipong-castcle) [inline apps/users/src/app/app.controller.ts:932]:
```suggestion
```

already validated in `FeedQuery`

File: libs/database/src/lib/dtos/common.dto.ts [hunk 3]
```diff
@@ -107,3 +114,20 @@ export class CastcleParticipate {
...
+  @TransformStringToArrayOfStrings()
+  @IsOptional()
+  @IsArray()
+  @IsEnum(UserField, { each: true })
+  userFields?: UserField[];
+
+  hasRelationshipExpansion = (() => {
+    return Boolean(this.userFields?.includes(UserField.Relationships));
+  }) as unknown as boolean;
+}
```
Comment (kittipong-castcle) [inline libs/database/src/lib/dtos/common.dto.ts:133]: UserField and ExpansionQuery have already moved to `feed.dto.ts`. I'm not sure that why we need to duplicate them in `common.dto.ts`.

File: libs/database/src/lib/dtos/feed.dto.ts [hunk 4]
```diff
@@ -21,39 +21,10 @@
...
-  @IsArray()
-  @IsEnum(UserField, { each: true })
-  userFields?: UserField[];
-
-  hasRelationshipExpansion = (() => {
-    return Boolean(this.userFields?.includes(UserField.Relationships));
-  }) as unknown as boolean;
-}
+import { IsMongoId, IsNumber, IsOptional, Max } from 'class-validator';
+import { DEFAULT_QUERY_OPTIONS, ExpansionQuery } from './common.dto';
```
Comment (kittipong-castcle) [inline libs/database/src/lib/dtos/feed.dto.ts:27]: Merging conflict?
Moving back ExpansionQuery and UserField to `common.dto` can cause circular dependency.

File: libs/database/src/lib/services/user.service.ts [hunk 5]
```diff
@@ -818,4 +833,98 @@ Message: ${message}`
...
+    relationUserId: string,
+    hasRelationshipExpansion: boolean
+  ) => {
+    return getRelationship(
+      relationships,
+      viewerId,
+      relationUserId,
+      hasRelationshipExpansion
+    );
+  };
```
Comment (kittipong-castcle) [inline libs/database/src/lib/services/user.service.ts:871]:
```suggestion
```

Same but different naming?

File: libs/database/src/lib/services/user.service.ts [hunk 6]
```diff
@@ -818,4 +833,98 @@ Message: ${message}`
...
+    const result: UserDocument[] = [];
+    if (accountReferee && accountReferee.length > 0) {
+      this.logger.log('Get user.');
+      Promise.all(
+        accountReferee.map(async (x) =>
+          result.push(await this.getUserFromAccountId(x.referringAccount._id))
+        )
+      );
+      this.logger.log('Success get referee.');
+    }
```
Comment (kittipong-castcle) [inline libs/database/src/lib/services/user.service.ts:924]:
```suggestion
    const result = await Promise.all(
        accountReferee?.map((x) => this.getUserFromAccountId(x.referringAccount._id) ?? []
    );
```

**Review by kittipong-castcle — CHANGES_REQUESTED** [review]

**Review by kittipong-castcle — APPROVED** [review]

**Pull request merged**